### PR TITLE
Typescript refactoring, seperate services, custom methods and vitepress docs

### LINKED ISSUE #144: Rework
I was thinking about whether we should rework the repo.
I have two suggestion in mind.

1. Moving the source code to a TypeScript base.
I would argue this is more maintainable and this would also help with intellisense in various IDE's.

2. This is more radical - Moving away from actions in `data` and instead use `authmanagement/{action}` .
So a reset password would move from 
```json
{
    "action": "resetShortToken",
    "data": { ... }
}
```
To a more REST like approach with `/authmanagement/reset` 
This would make it work more like a REST API and as well as a standard feathers service.

I realize this would be breaking changes, but I think it would benefit the overall structure of calling the API

Let me know what you think, and if I should elaborate.

### PR DESCRIPTION
**Pull request by fratzinger**
# tl;dr

Checkout the migration guide: https://feathers-a-m.netlify.app/migration.html

## New Features:
- refactored to typescript. There shouldn't be a breaking change, though!
- new docs: https://feathers-a-m.netlify.app/
- each `action` has a new separate service now:

see #159 for the beginning of Discussion. We continue the refactoring here.

### Todos:
- [x] Port `src` to typescript is pretty much done with types
  - [x] Typescript 
  - [x] propper typings
  - [x] service class
  - [ ] make sure `client` is backward compatible
- [x] Port `test` to typescript
  - [x] Typescript
  - [x] propper typings
- [x] seperate services/routes as suggested in #144 
  - [x] seperate Class with create
  - [x] call services create methods in main service for backward compatibility
  - [x] one service for every `action`
  - [x] use `_create` for every service to make it possible to skip hooks as for `adapter-commons`
  - [x] make sure `service.publish` does not get hit to prevent channel emit
  - [x] add custom methods (for feathers^5)
  - [x] register services under route (`app.use`)
  - [x] add more tests for this
- [x] add docs
  - [x] VitePress
  - [x] Go through everything

### New Stuff:
- added option: `passwordField` for dynamic passwordField. It was a fixed variable before
- added services 
- can call `_create` function on all services to bypass hooks
- separate services for each action (except `action: 'options'`)
- each method as hookless custom method (e.g.: `app.service('authManagement')._resetPasswordLong(...)`)
- each method as custom method (e.g.: `app.service('authManagement').resetPasswordLong(...)`
- recommended use of `app.use('auth-management, new AuthenticationManagementService(app, { ... })` instead of `app.configure(authManagement)`

### Todo:
- [ ] Update links to source code in service-hooks.md.
- [x] Removal of old docs + update README
- [x] Add logo
- [x] add netlify page
- [x] prerelease

### IDE-Support:
image

closes #144, #138

### CODE REVIEW

**Comment by fratzinger** [comment]
This took way too long. But I think we're close to a first `pre` release.
@OnnoGabriel helped me a lot with the docs! Thanks to him, we're able to show the first deploy of the new `feathers-authentication-management` docs: https://heuristic-euler-f28fa9.netlify.app/ (temp preview deploy).

Now it's time to decide what this is and where it should live. Should it be: 
- www.feathers-authentication-management.com
- feathers-authentication-management.netlify.app
- f-a-m.netlify.app
- https://auth-management.feathersjs.com/

What do you think? @OnnoGabriel, @claustres, @TheSinding, @daffl

**Comment by claustres** [comment]
First of all thanks for this great job ! About the docs domain select what is the more simple to manage for you. Maybe using github pages and the domain they provided by github is simple enough for maintainers to avoid playing with DNS ? 

As far as I can see this should be backward compatible, this is the cherry on the cake.

**Comment by fratzinger** [comment]
Currently I only have knowledge about netlify. It's pretty simple. I host docs of `feathers-casl` and `feathers-trigger` with it:
- `feathers-casl` [github repo](https://github.com/fratzinger/feathers-casl) - [feathers-casl.netlify.app](https://feathers-casl.netlify.app/)
- `feathers-trigger` [github repo](https://github.com/fratzinger/feathers-trigger) - [feathers-tigger.netlify.app](https://feathers-trigger.netlify.app/)

But maybe it should be treated like https://hooks-common.feathersjs.com/ or maybe even a more curated ecosystem like https://github.com/feathers-plus/docs was with links to several packages in the header.

**Comment by claustres** [comment]
The advantage of using Github pages is that any maintainer having access to the repository can update the doc when committing, see [vitepres docs](https://vitepress.vuejs.org/guide/deploy.html#github-pages). Docs are automatically hosted/updated on repository.github.io domain.

I don't use Netlify but relying on an external service often requires to manage accounts/rights on this service as well for package maintainers in order to be able to deploy/update the docs. But maybe you have already setup an automated process to do so when committing ? That was just a suggestion anyway, in a first step do what is easier for you, then we can iterate.

Thanks again for sharing all this work @fratzinger @OnnoGabriel.

**Comment by fratzinger** [comment]
Thanks for the acknowledgement.

netlify totally got this. I guess gh pages, netlify and vercel are the same for open source. Automatically accessible deploy previews like here. It's really three clicks of setup.
image

> manage accounts/rights

That's a point though!

**Comment by TheSinding** [comment]
This is awesome work @fratzinger & @OnnoGabriel !

I'll have to agree with @claustres - I think GH Pages would be the better option, purely from the standpoint that it would be easier for maintainers coming and leaving the project to edit it, also nobody would have to manage the domain.

That being said, I think the best domain would be to attach us to the "main" domain [feathersjs.com](feathersjs.com), if thats not possible, I think the best way would be GH pages

**Comment by fratzinger** [comment]
The first pre-release is ready and looks for early adopters:
```
npm i feathers-authentication-management@pre
```
migration should be fairly easy. Everyone please let me know if you tried it.

**Comment by OnnoGabriel** [comment]
Great work, @fratzinger! I tested `feathers-authentication-management@pre` in a Feathers v4 app with an notifier function sending e-mails. All tested actions run well: `resendVerifySignup`, `verifySignupLong`, `verifySignupSetPasswordLong`, `sendResetPwd`, `resetPwdLong`, `identityChange` (I did not test the short token actions).

**Review by stevek-pro — COMMENTED** [review]
Tested and works

**Comment by stevek-pro** [comment]
I tested verifySignup functionality and it works as advertised. Well done. Will check the other hooks this week.

**Comment by stevek-pro** [comment]
Mongoose Models need `type: String` for `verifyChanges` instead of `type: Array`. This should be mentioned in docs in my opinion to safe others many troubleshooting hours.
